cmd/audit/audit.go

```diff
@@ -81,45 +82,60 @@ func read() error {
 	}
 
 	log.Debugln("starting netlink client")
-	client, err := libaudit.NewAuditClient(diagWriter)
-	if err != nil {
-		return err
-	}
 
-	status, err := client.GetStatus()
-	if err != nil {
-		return errors.Wrap(err, "failed to get audit status")
-	}
-	log.Infof("received audit status=%+v", status)
+	var err error
+	var client *libaudit.AuditClient
+	if *receiveOnly {
+		client, err = libaudit.NewMulticastAuditClient(diagWriter)
+		if err != nil {
+			return errors.Wrap(err, "failed to create receive-only audit client")
+		}
+	} else {
+		client, err = libaudit.NewAuditClient(diagWriter)
+		if err != nil {
+			return errors.Wrap(err, "failed to create audit client")
+		}
 
-	if status.Enabled == 0 {
-		log.Debugln("enabling auditing in the kernel")
-		if err = client.SetEnabled(true, libaudit.WaitForReply); err != nil {
-			return errors.Wrap(err, "failed to set enabled=true")
+		status, err := client.GetStatus()
+		if err != nil {
+			return errors.Wrap(err, "failed to get audit status")
 		}
-	}
+		log.Infof("received audit status=%+v", status)
 
-	if status.RateLimit != uint32(*rate) {
-		log.Debugf("setting rate limit in kernel to %v", *rate)
-		if err = client.SetRateLimit(uint32(*rate), libaudit.NoWait); err != nil {
-			return errors.Wrap(err, "failed to set rate limit to unlimited")
+		if status.Enabled == 0 {
+			log.Debugln("enabling auditing in the kernel")
+			if err = client.SetEnabled(true, libaudit.WaitForReply); err != nil {
+				return errors.Wrap(err, "failed to set enabled=true")
+			}
 		}
-	}
 
-	if status.BacklogLimit != uint32(*backlog) {
-		log.Debugf("setting backlog limit in kernel to %v", *backlog)
-		if err = client.SetBacklogLimit(uint32(*backlog), libaudit.NoWait); err != nil {
-			return errors.Wrap(err, "failed to set backlog limit")
+		if status.RateLimit != uint32(*rate) {
+			log.Debugf("setting rate limit in kernel to %v", *rate)
+			if err = client.SetRateLimit(uint32(*rate), libaudit.NoWait); err != nil {
+				return errors.Wrap(err, "failed to set rate limit to unlimited")
+			}
+		}
+
+		if status.BacklogLimit != uint32(*backlog) {
+			log.Debugf("setting backlog limit in kernel to %v", *backlog)
+			if err = client.SetBacklogLimit(uint32(*backlog), libaudit.NoWait); err != nil {
+				return errors.Wrap(err, "failed to set backlog limit")
+			}
 		}
-	}
 
-	log.Debugf("sending message to kernel registering our PID (%v) as the audit daemon", os.Getpid())
-	if err = client.SetPID(libaudit.NoWait); err != nil {
-		return errors.Wrap(err, "failed to set audit PID")
+		log.Debugf("sending message to kernel registering our PID (%v) as the audit daemon", os.Getpid())
+		if err = client.SetPID(libaudit.NoWait); err != nil {
+			return errors.Wrap(err, "failed to set audit PID")
+		}
 	}
+	defer client.Close()
```

Comment:
Should this defer also happen on some of the errors above?

Reply:
Fixed